### PR TITLE
Code restructuring for Multi-Stage Curriculum Learning

### PR DESCRIPTION
**Pull request by RealLast**
# *Code restructuring for Multi-Stage Curriculum Learning

## :recycle: Current situation & Problem
* Code, tests and data were all scattered across main and subfolders
* Data cleaning difficult, as it was under time_series_data/raw_data, mixed with the code
* Structure did not follow recent best practices for Deep Learning repositories 
* Previous code base required to implement training twice, once for EmbedHealthSP and once for EmbedHealthFlamingo 

## :gear: Release Notes
*Restructured code base to adhere to the following structure:
- src
   - Model
      - EmbedHealthSP  
      - EmbedHealthFlamingo 
- data (contains downloaded raw data, can be deleted for cleaning)
   - M4
   - Sleep
   - Pamap
   - ...   
- tests
   - unit tests for testing m4 loader, testing training of EmbedHealthSP and EmbedHealthFlamingo separately, etc. 
- notebooks   

* Updated README
* Moved M4 Time Series Caption Dataset outside of this repository into a standalone repo: [M4TSCaptioning](https://github.com/StanfordBDHG/M4TimeSeriesCaptionDataset)
* Implemented DDP-aware Multi-Stage curriculum learning into a single code
  * Stage 1: Multiple Choice Question Answering (MCQA) on TSQA
  * Stage 2: Time Series Captioning on M4
  * Stages are trained sequentially, reusing best model of previous stage 
  * Same training loops are reused for EmbedHealthSP and EmbedHealthFlamingo
  * Verbose debug statements, informations 
  
  * DDP support available for both EmbedHealthSP and EmbedHealthFlamingo  

* Detailed metric tracking and checkpointing; best models of each stage are stored in separate folders per model, test evaluation metrics are stored as well   
```
results/
├── EmbedHealthSP/
│   ├── stage1_mcq/
│   │   ├── checkpoints/
│   │   │   └── best_model.pt
│   │   └── results/
│   │       ├── test_predictions.jsonl
│   │       └── metrics.json
│   ├── stage2_captioning/
│   │   ├── checkpoints/
│   │   │   └── best_model.pt
│   │   └── results/
│   │       ├── test_predictions.jsonl
│   │       └── metrics.json
│   └── curriculum_results.json
└── EmbedHealthFlamingo/
    ├── stage1_mcq/
    │   ├── checkpoints/
    │   │   └── best_model.pt
    │   └── results/
    │       ├── test_predictions.jsonl
    │       └── metrics.json
    ├── stage2_captioning/
    │   ├── checkpoints/
    │   │   └── best_model.pt
    │   └── results/
    │       ├── test_predictions.jsonl
    │       └── metrics.json
    └── curriculum_results.json
```

## :books: Documentation
*Please ensure that you properly document any additions in conformance to [Spezi Documentation Guide](https://github.com/StanfordSpezi/.github/blob/main/DOCUMENTATIONGUIDE.md).*
*You can use this section to describe your solution, but we encourage contributors to document your reasoning and changes using in-line documentation.* 

## :white_check_mark: Testing
*Please ensure that the PR meets the testing requirements set by CodeCov and that new functionality is appropriately tested.*
*This section describes important information about the tests and why some elements might not be testable.*

### CODE REVIEW

File: curriculum_learning.py [hunk 1]
```diff
@@ -0,0 +1,1085 @@
...
+        print(f"   Epochs: {num_epochs}")
+        if self.model_type == "EmbedHealthSP":
+            print(f"   Encoder LR: {lr_encoder:.2e}")
+            print(f"   Projector LR: {lr_projector:.2e}")
+        else:
+            print(f"   Base LR: {lr_base:.2e}")
+        print(f"   Batch size per GPU: {batch_size}")
+        if self.world_size > 1:
+            print(f"   Effective batch size: {batch_size * self.world_size}")
+        print()
```
[the remark this answers is not in the document]
Comment (RealLast) [reply curriculum_learning.py:558]: Right, thank you!

File: test/test_curriculum_trainer.py
[the remark this answers is not in the document]
Comment (RealLast) [reply test/test_curriculum_trainer.py:1]: :rocket: :rocket:

File: src/time_series_datasets/ecg_qa/ecgqa_loader.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+
```
[the remark this answers is not in the document]
Comment (RealLast) [reply src/time_series_datasets/ecg_qa/ecgqa_loader.py:1]: Thanks for catching this! Not meant for release yet, gonna be a separate PR.

File: curriculum_learning.py [hunk 3]
```diff
@@ -0,0 +1,1085 @@
...
+                 dist_url: str = "env://", dist_backend: str = "nccl", local_rank: int = int(os.environ.get("LOCAL_RANK", 0)),
+                 llm_id: str = None):
+        """
+        Initialize the curriculum trainer.
+        
+        Args:
+            model_type: Either 'EmbedHealthSP' or 'EmbedHealthFlamingo'
+            device: Device to use for training ('cuda', 'mps', or 'cpu')
+            gradient_checkpointing: Enable gradient checkpointing
+            dist_url: URL used to set up distributed training
```
[the remark this answers is not in the document]
Comment (RealLast) [reply curriculum_learning.py:80]: I think its the default for DDP, we had that before as well. torchrun starts one process per GPU and then they communicate via local sockets as far as I understood.
